google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -477,28 +767,66 @@
   @Nonnull
   public ApiFuture<Void> flush() {
     verifyNotClosed();
-    final SettableApiFuture<Void> flushComplete = SettableApiFuture.create();
-    List<SettableApiFuture<WriteResult>> writeFutures = new ArrayList<>();
+    final ArrayList<ApiFuture<Void>> pendingOperationsAtFlush =
+        Lists.newArrayList(pendingOperations);
+    return ApiFutures.transformAsync(
+        ApiFutures.immediateFuture(null),
+        new ApiAsyncFunction<Object, Void>() {
+          public ApiFuture<Void> apply(Object o) throws Exception {
+            return performFlush(pendingOperationsAtFlush);
+          }
+        },
+        bulkWriterExecutor);
+  }
+
+  private ApiFuture<Void> performFlush(final List<ApiFuture<Void>> pendingOperations) {
     for (BulkCommitBatch batch : batchQueue) {
       batch.markReadyToSend();
-      writeFutures.addAll(batch.getPendingFutures());
     }
-    sendReadyBatches();
-    ApiFutures.successfulAsList(writeFutures)
+
+    // Send all scheduled operations on the BatchQueue first.
+    sendReadyBatches(batchQueue);
+    final SettableApiFuture<Void> batchQueueFlushComplete = SettableApiFuture.create();
+    final SettableApiFuture<Void> flushComplete = SettableApiFuture.create();
+    ApiFutures.successfulAsList(pendingBatches)
         .addListener(
             new Runnable() {
               public void run() {
-                flushComplete.set(null);
+                batchQueueFlushComplete.set(null);
               }
             },
             MoreExecutors.directExecutor());
+
+    // Afterwards, send all accumulated retry operations. Wait until the retryBatchQueue is cleared.
+    // This way, operations scheduled after flush() will not be sent until the retries are done.
+    batchQueueFlushComplete.addListener(
+        new Runnable() {
+          public void run() {
+            if (retryBatchQueue.size() > 0) {
+              for (BulkCommitBatch batch : retryBatchQueue) {
+                batch.markReadyToSend();
+              }
+              sendReadyBatches(retryBatchQueue);
+            }
+            ApiFutures.successfulAsList(pendingOperations)
```

Comment:
I think this doesn't need to be chained here. Could we move it outside of this block to reduce chaining a bit?